### PR TITLE
feat: cosmwasmpool core logic and tests

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

This is the core logic and full integration of `x/cosmwasmpool` into the rest of the system.

It is tested with the version of transmuter contract as of `main` branch here:  https://github.com/osmosis-labs/transmuter

Spec: https://github.com/osmosis-labs/osmosis/tree/main/x/cosmwasmpool

The remaining steps are:
- implement gov props per: https://github.com/osmosis-labs/osmosis/pull/5346
- add queries and CLI
- holistically test transmuter with tokenfactory

## Testing and Verifying

This change added tests

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [x] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [x] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/cosmwasmpool/keeper.go [hunk 1]
```diff
@@ -43,3 +42,32 @@ func (k Keeper) GetParams(ctx sdk.Context) (params types.Params) {
...
+	k.contractKeeper = contractKeeper
+}
+
+// Set the wasm keeper.
+func (k *Keeper) SetWasmKeeper(wasmKeeper types.WasmKeeper) {
+	k.wasmKeeper = wasmKeeper
+}
+
+// convertToCosmwasmPool converts a poolI to a CosmWasmExtension.
+func (k *Keeper) convertToCosmwasmPool(poolI poolmanagertypes.PoolI) (types.CosmWasmExtension, error) {
```
Comment (nicolaslara) [inline x/cosmwasmpool/keeper.go:62]: nit: convert makes it sounds like you're modifying the pool, but we're just casting it to the new type. Maybe rename to something like `asCosmwasmPool()`?

File: x/cosmwasmpool/pool_module.go [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+//
+// Parameters:
+// - ctx: The SDK context.
+// - poolId: The unique identifier of the pool.
+//
+// Returns:
+// - poolmanagertypes.PoolI: The pool interface of the corresponding pool model, if found.
+// - error: An error if the pool model is not found; otherwise, nil.
+func (k Keeper) GetPool(ctx sdk.Context, poolId uint64) (poolmanagertypes.PoolI, error) {
+	concentratedPool, err := k.GetPoolById(ctx, poolId)
```
Comment (nicolaslara) [inline x/cosmwasmpool/pool_module.go:66]: nit: rename concentratedPool to pool (as this is any PoolI)

File: x/cosmwasmpool/pool_module.go [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+//
+// Parameters:
+// - ctx: The current SDK Context used to access the store.
+//
+// Returns:
+//   - A slice of PoolI interfaces if the operation is successful. Each element in the slice
+//     represents a pool that was stored in the keeper.
+//   - An error if unmarshalling fails for any of the values fetched from the store.
+//     In this case, the slice of PoolI interfaces will be nil.
+func (k Keeper) GetPools(ctx sdk.Context) ([]poolmanagertypes.PoolI, error) {
```
Comment (nicolaslara) [inline x/cosmwasmpool/pool_module.go:88]: who can call this? I'm concerned about performance as this is a linear operation

File: x/cosmwasmpool/keeper.go (hunk 1 above)
Comment (p0mvn) [reply x/cosmwasmpool/keeper.go]: renamed as well as relevant helpers in other modules

File: x/cosmwasmpool/pool_module.go [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+// - denoms: A slice of strings representing the asset denoms in the liquidity pool.
+// - err: An error if the pool cannot be found or if the CosmWasm query fails.
+func (k Keeper) GetPoolDenoms(ctx sdk.Context, poolId uint64) (denoms []string, err error) {
+	cosmwasmPool, err := k.GetPoolById(ctx, poolId)
+	if err != nil {
+		return nil, err
+	}
+
+	liquidity := cosmwasmPool.GetTotalPoolLiquidity(ctx)
+	if liquidity.Len() < 2 {
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/cosmwasmpool/pool_module.go]: Agreed, removed this

**Comment by p0mvn** [comment]
Addressed all comments. Going to merge once CI passes. Thanks @nicolaslara

**Pull request merged**